modules/aggregation-layers/src/grid-aggregation-layer.js

```diff
@@ -20,22 +20,315 @@
 
 import AggregationLayer from './aggregation-layer';
 import GPUGridAggregator from './utils/gpu-grid-aggregation/gpu-grid-aggregator';
+import {AGGREGATION_OPERATION} from './utils/aggregation-operation-utils';
+import {Buffer} from '@luma.gl/core';
+import {WebMercatorViewport, log, COORDINATE_SYSTEM} from '@deck.gl/core';
+import GL from '@luma.gl/constants';
+import {Matrix4} from 'math.gl';
+import {getBoundingBox} from './utils/grid-aggregation-utils';
+import {getValueFunc} from './utils/aggregation-operation-utils';
+import BinSorter from './utils/bin-sorter';
+import {pointToDensityGridDataCPU, getGridOffset} from './cpu-grid-layer/grid-aggregator';
 
 export default class GridAggregationLayer extends AggregationLayer {
-  initializeState(aggregationProps) {
+  initializeState({aggregationProps, getCellSize}) {
     const {gl} = this.context;
     super.initializeState(aggregationProps);
     this.setState({
-      gpuGridAggregator: new GPUGridAggregator(gl, {id: `${this.id}-gpu-aggregator`})
+      // CPU aggregation results
+      layerData: {},
+      gpuGridAggregator: new GPUGridAggregator(gl, {id: `${this.id}-gpu-aggregator`}),
+      cpuGridAggregator: pointToDensityGridDataCPU
     });
   }
 
+  updateState(opts) {
+    // will get new attributes
+    super.updateState(opts);
+
+    // update bounding box and cellSize
+    this._updateAggregationState(opts);
+
+    let aggregationDirty = false;
+    const {needsReProjection, gpuAggregation} = this.state;
+    const needsReAggregation = this._isAggregationDirty(opts);
+    if (this.getNumInstances() <= 0) {
+      return;
+    }
+    if (needsReProjection || (gpuAggregation && needsReAggregation)) {
+      this._updateAccessors(opts);
+      this._resetResults();
+      this._getAggregatedData(opts);
+      aggregationDirty = true;
+    }
+    if (!gpuAggregation && (aggregationDirty || needsReAggregation)) {
+      // In case of CPU aggregation
+      this._resetResults();
+      this._updateWeightBins();
+      this._uploadAggregationResults();
+      aggregationDirty = true;
+    }
+
+    this.setState({aggregationDirty});
+  }
+
   finalizeState() {
-    super.finalizeState();
+    const {count} = this.state.weights;
+    if (count && count.aggregationBuffer) {
+      count.aggregationBuffer.delete();
+    }
     const {gpuGridAggregator} = this.state;
     if (gpuGridAggregator) {
       gpuGridAggregator.delete();
     }
+    super.finalizeState();
+  }
+
+  // Private
+
+  _updateAggregationState(opts) {
```

Comment:
Why do we need this method? Just use the two methods below inline?

Reply:
Will remove it, had other code before I made the final PR.